<h2> Using KNieghbors, DecisionTree, and Random Forest to predict a U Prep student's happiness, stress, and energy level at any given time. 

<h4> by Tessa de Vries


Intro: For my preliminary round of classifying I used the following 6 features for each level I was trying to predict. Day of the week, # of graded assessements today, hours slept last night, # of times exercised in the last 7 days (20+ min), and the two levels that I was not trying to predict.  These were selected because in my personal experience they are the factors that are most influential in the output levels. 

<h3> ACCURACY: Predicting Stress Level (6 features- DecisionTree)

In [60]:
from sklearn import tree
import numpy as np
import random 

raw_data = open("stress.csv")
dataset = np.loadtxt(raw_data, delimiter=",")
X = dataset[:,0:6]
y = dataset[:,6]

from sklearn import datasets
import random
import statistics

from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = .20)

from sklearn import tree
tree_classifier = tree.DecisionTreeClassifier();
tree_classifier = tree_classifier.fit(X_train, y_train)
tree_predictions = tree_classifier.predict(X_test)

from sklearn.metrics import accuracy_score
print(accuracy_score(tree_predictions, y_test))

0.115384615385


<h3> ACCURACY: Predicting Stresst Level (6 features- KNeighbors)

In [61]:
from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = .15)

from sklearn import neighbors
knn_classifier = neighbors.KNeighborsClassifier()
knn_classifier = knn_classifier.fit(X_train, y_train)
knn_predictions = knn_classifier.predict(X_test)

from sklearn.metrics import accuracy_score
print(accuracy_score(knn_predictions, y_test))

0.051724137931


<h3> ACCURACY: Predicting Stresst Level (6 features- RandomForest)

In [62]:
from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = .15)

from sklearn import ensemble
rf_classifier = ensemble.RandomForestClassifier()
rf_classifier = rf_classifier.fit(X_train, y_train)
rf_predictions = rf_classifier.predict(X_test)

from sklearn.metrics import accuracy_score
print(accuracy_score(rf_predictions, y_test))


0.0172413793103


<h3> ACCURACY: Predicting Happiness Level (6 features-DecisionTree)

In [56]:
from sklearn import tree
import numpy as np
import random 

raw_data = open("happy1.csv")
dataset = np.loadtxt(raw_data, delimiter=",")
X = dataset[:,0:6]
y = dataset[:,6]


In [57]:
from sklearn import datasets
import random
import statistics

from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = .15)

from sklearn import tree
tree_classifier = tree.DecisionTreeClassifier();
tree_classifier = tree_classifier.fit(X_train, y_train)
tree_predictions = tree_classifier.predict(X_test)

from sklearn.metrics import accuracy_score
print(accuracy_score(tree_predictions, y_test))
       
       

0.275862068966


<h3> ACCURACY: Predicting Happiness (6 features-KNieghbors)

In [58]:
from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = .20)

from sklearn import neighbors
knn_classifier = neighbors.KNeighborsClassifier()
knn_classifier = knn_classifier.fit(X_train, y_train)
knn_predictions = knn_classifier.predict(X_test)

from sklearn.metrics import accuracy_score
print(accuracy_score(knn_predictions, y_test))

0.128205128205


<h3> ACCURACY: Predicting Happiness (6 features- RandomForest) 

In [59]:
from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = .20)

from sklearn import ensemble
rf_classifier = ensemble.RandomForestClassifier()
rf_classifier = rf_classifier.fit(X_train, y_train)
rf_predictions = rf_classifier.predict(X_test)

from sklearn.metrics import accuracy_score
print(accuracy_score(rf_predictions, y_test))




0.230769230769


<h3> ACCURACY: Predicting Energy Level (6 features- DecisionTree)

In [63]:
from sklearn import tree
import numpy as np
import random 

raw_data = open("energy.csv")
dataset = np.loadtxt(raw_data, delimiter=",")
X = dataset[:,0:6]
y = dataset[:,6]

from sklearn import datasets
import random
import statistics

from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = .15)

from sklearn import tree
tree_classifier = tree.DecisionTreeClassifier();
tree_classifier = tree_classifier.fit(X_train, y_train)
tree_predictions = tree_classifier.predict(X_test)

from sklearn.metrics import accuracy_score
print(accuracy_score(tree_predictions, y_test))

0.120689655172


<h3> ACCURACY: Predicting Happiness( 6 features- KNieghbors)

In [64]:
from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = .15)

from sklearn import neighbors
knn_classifier = neighbors.KNeighborsClassifier()
knn_classifier = knn_classifier.fit(X_train, y_train)
knn_predictions = knn_classifier.predict(X_test)

from sklearn.metrics import accuracy_score
print(accuracy_score(knn_predictions, y_test))

0.172413793103


<h3> ACCURACY: Predicting Happiness (6 fetaures- RandomForest)

In [65]:
from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = .15)

from sklearn import ensemble
rf_classifier = ensemble.RandomForestClassifier()
rf_classifier = rf_classifier.fit(X_train, y_train)
rf_predictions = rf_classifier.predict(X_test)

from sklearn.metrics import accuracy_score
print(accuracy_score(rf_predictions, y_test))


0.172413793103


Analysis: Due to the fact that my accuracies were fairly low across the board, I decided to do further analysis. Initially I did this through using regression (I uploaded the file with this code--it's name is "Final (regressions)") but that did not provide any better results. I was not able to find any strong correlations between any single feature and my intended target level (happiness, stress, or energy). I decided to go back to machine learning, specifically the three techniques I used previously (Decision Tree, KNieghbors, and RandomForest). I thought that reducing the features from 6 to 3 for each level would produce higher accuracy levels. I also selected features that I thought were the most influential in determining the outputs. The features for Happiness were : day of the week, hours of sleep last night, and stress level. The features for Energy were: happiness level, # of graded assessements today, and hours of sleep last night. The features for Stress were: # of graded assessements today, # of times exercised in the last 7 days (20+ min), and hours of sleep last night. 


<h3> ACCURACY: Predicting Stress Level(3 features- DecisionTree, KNieghbors, RandomForest)

In [4]:
from sklearn import tree
import numpy as np
import random 

raw_data = open("stress1.csv")
dataset = np.loadtxt(raw_data, delimiter=",")
X = dataset[:,0:3]
y = dataset[:,3]

from sklearn import datasets
import random
import statistics

from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = .20)

from sklearn import tree
tree_classifier = tree.DecisionTreeClassifier();
tree_classifier = tree_classifier.fit(X_train, y_train)
tree_predictions = tree_classifier.predict(X_test)

from sklearn.metrics import accuracy_score
print(accuracy_score(tree_predictions, y_test))

from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = .15)

from sklearn import neighbors
knn_classifier = neighbors.KNeighborsClassifier()
knn_classifier = knn_classifier.fit(X_train, y_train)
knn_predictions = knn_classifier.predict(X_test)

from sklearn.metrics import accuracy_score
print(accuracy_score(knn_predictions, y_test))

from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = .15)

from sklearn import ensemble
rf_classifier = ensemble.RandomForestClassifier()
rf_classifier = rf_classifier.fit(X_train, y_train)
rf_predictions = rf_classifier.predict(X_test)

from sklearn.metrics import accuracy_score
print(accuracy_score(rf_predictions, y_test))



0.166666666667
0.120689655172
0.103448275862


<h3> ACCURACY: Predicting Energy Level(3 features- DecisionTree, KNieghbors, RandomForest)

In [7]:
from sklearn import tree
import numpy as np
import random 

raw_data = open("energy1.csv")
dataset = np.loadtxt(raw_data, delimiter=",")
X = dataset[:,0:3]
y = dataset[:,3]

from sklearn import datasets
import random
import statistics

from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = .20)

from sklearn import tree
tree_classifier = tree.DecisionTreeClassifier();
tree_classifier = tree_classifier.fit(X_train, y_train)
tree_predictions = tree_classifier.predict(X_test)

from sklearn.metrics import accuracy_score
print(accuracy_score(tree_predictions, y_test))

from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = .15)

from sklearn import neighbors
knn_classifier = neighbors.KNeighborsClassifier()
knn_classifier = knn_classifier.fit(X_train, y_train)
knn_predictions = knn_classifier.predict(X_test)

from sklearn.metrics import accuracy_score
print(accuracy_score(knn_predictions, y_test))

from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = .15)

from sklearn import ensemble
rf_classifier = ensemble.RandomForestClassifier()
rf_classifier = rf_classifier.fit(X_train, y_train)
rf_predictions = rf_classifier.predict(X_test)

from sklearn.metrics import accuracy_score
print(accuracy_score(rf_predictions, y_test))


0.166666666667
0.155172413793
0.137931034483


<h3> ACCURACY: Predicting Happiness Level(3 features- DecisionTree, KNieghbors, RandomForest)

In [8]:
from sklearn import tree
import numpy as np
import random 

raw_data = open("happy2.csv")
dataset = np.loadtxt(raw_data, delimiter=",")
X = dataset[:,0:3]
y = dataset[:,3]

from sklearn import datasets
import random
import statistics

from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = .20)

from sklearn import tree
tree_classifier = tree.DecisionTreeClassifier();
tree_classifier = tree_classifier.fit(X_train, y_train)
tree_predictions = tree_classifier.predict(X_test)

from sklearn.metrics import accuracy_score
print(accuracy_score(tree_predictions, y_test))

from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = .15)

from sklearn import neighbors
knn_classifier = neighbors.KNeighborsClassifier()
knn_classifier = knn_classifier.fit(X_train, y_train)
knn_predictions = knn_classifier.predict(X_test)

from sklearn.metrics import accuracy_score
print(accuracy_score(knn_predictions, y_test))

from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = .15)

from sklearn import ensemble
rf_classifier = ensemble.RandomForestClassifier()
rf_classifier = rf_classifier.fit(X_train, y_train)
rf_predictions = rf_classifier.predict(X_test)

from sklearn.metrics import accuracy_score
print(accuracy_score(rf_predictions, y_test))



0.153846153846
0.137931034483
0.224137931034


Unfortunately my accuracies were not exceptionally higher or even close to what I wanted (around 80%). However, the accuracy for stress almost doubled from 6 to 13%! And the accuracy for energy increased significantly as well. Happiness on the other hand decreased about 4%. 

After doing this analysis I realize how personal and arbitrary "happiness, stress, and energy" levels are. One person's 7 may be another person's 3 (so to speak). This also makes me question how thoroughly and thoughtfully people fill out surveys. Perhaps the reason my accuracy levels were so low was because the data wasn't accurate do to peoples' inaccuracy in the survey. I have a feeling that people may have started to do the survey rushed considering it was for a Stats Class and required every week for several weeks. 

I would have enjoyed being able to code a model with a very high accuracy but I realize that you don't always get the results you initally desire. Furthermore, just because I didn't get very functional product not only doesn't mean I didn't put a lot of work into the lab but it also doesn't mean I didn't learn anything from the low accuracies. 